craft_parts/sources/errors.py

```diff
@@ -108,3 +109,26 @@ def __init__(self, source: str):
         resolution = "Make sure the source path is correct and accessible."
 
         super().__init__(brief=brief, resolution=resolution)
+
+
+class InvalidSnapPackage(SourceError):
+    """A snap package is invalid."""
+
+    def __init__(self, snap_file: str):
+        self.snap_file = snap_file
+        brief = f"Snap {snap_file!r} does not contain valid data."
+        resolution = "Ensure the source lists a proper snap file."
+
+        super().__init__(brief=brief, resolution=resolution)
+
+
+class PullError(SourceError):
+    """Failed pulling source."""
+
+    def __init__(self, *, command: List[str], exit_code: int):
+        self.command = command
+        self.exit_code = exit_code
+        cmd = " ".join(shlex.quote(i) for i in command)
+        brief = f"Failed to pull source: command {cmd!r} exited with code {exit_code}."
```

Comment:
`shlex.join(command)` is now available to avoid the above " ".join/quote op.

Personally I prefer this because it prints a command I can paste into the shell if I intend to reproduce.  Out of curiosity, what details could be hidden @facundobatista ?

Reply:
Also speaking from a lot of experience lately trying to reproduce commands being executed into provided environments :rofl:

Reply:
I don't know if the original author had any special reason not to do so other than humanize the output. Stringifying it hides things like `["ls -l"]` being used instead of `["ls", "-l"]`, so I agree that it's better to display it raw.